gobblin-core/src/main/java/org/apache/gobblin/converter/parquet/JsonElementConversionFactory.java

```diff
@@ -0,0 +1,484 @@
+import static org.apache.gobblin.converter.parquet.JsonSchema.*;
+import static org.apache.gobblin.converter.parquet.JsonSchema.InputType.STRING;
+import static parquet.schema.OriginalType.UTF8;
+import static parquet.schema.PrimitiveType.PrimitiveTypeName.BINARY;
+import static parquet.schema.PrimitiveType.PrimitiveTypeName.INT32;
+import static parquet.schema.PrimitiveType.PrimitiveTypeName.INT64;
+import static parquet.schema.Type.Repetition.REPEATED;
+
+
+/**
+ * <p>
+ * Creates a JsonElement to Parquet converter for all supported data types.
+ * </p>
+ *
+ * @author tilakpatidar
+ *
+ */
+public class JsonElementConversionFactory {
+
+  /**
+   * Use to create a converter for a single field from a parquetSchema.
+   *
+   * @param schema
+   * @param state
+   * @param repeated - Is the {@link Type} repeated in the parent {@link Group}
+   * @return
+   */
+  public static JsonElementConverter getConverter(JsonSchema schema, WorkUnitState state, boolean repeated) {
+
+    InputType fieldType = schema.getInputType();
+    switch (fieldType) {
+      case INT:
+        return new IntConverter(schema, repeated, state);
+
+      case LONG:
+        return new LongConverter(schema, repeated, state);
+
+      case FLOAT:
+        return new FloatConverter(schema, repeated, state);
+
+      case DOUBLE:
+        return new DoubleConverter(schema, repeated, state);
+
+      case BOOLEAN:
+        return new BooleanConverter(schema, repeated, state);
+
+      case STRING:
+        return new StringConverter(schema, repeated, state);
+
+      case ARRAY:
+        return new ArrayConverter(schema, state);
+
+      case ENUM:
+        return new EnumConverter(schema, state);
+
+      case RECORD:
+        return new RecordConverter(schema, state);
+
+      case MAP:
+        return new MapConverter(schema, state);
+
+      default:
+        throw new UnsupportedOperationException(fieldType + " is unsupported");
+    }
+  }
+
+  /**
+   * Converts a JsonElement into a supported ParquetType
+   * @author tilakpatidar
+   *
+   */
+  public static abstract class JsonElementConverter {
+    protected final JsonSchema jsonSchema;
+    protected final WorkUnitState state;
+    protected Type schema;
+
+    protected JsonElementConverter(JsonSchema schema, WorkUnitState state) {
+      this.jsonSchema = schema;
+      this.state = state;
+    }
+
+    /**
+     * Convert value to a parquet type and perform null check.
+     * @param value
+     * @return Parquet safe type
+     */
+    public Object convert(JsonElement value) {
+      if (value.isJsonNull()) {
+        if (this.jsonSchema.isNullable()) {
+          return null;
+        }
+        throw new RuntimeException(
+            "Field: " + this.jsonSchema.getColumnName() + " is not nullable and contains a null value");
+      }
+      return convertField(value);
+    }
+
+    /**
+     * Returns a {@link Type} parquet schema
+     * @return
+     */
+    public Type schema() {
```

Comment:
This one is better to be an abstract method since `schema` can only be set in subclasses. Also remove the `schema` property from this class.